RISC-V: Round up pre-populated stack frame to arch stack alignment

Teng-Shih-Wei opened the pull request:
The stack frame size, used for context switch, is rounded up to 16-bytes
alignment. Therefore, we need round down the pointer of top of the
pre-populated stack frame so that the preserved stack frame size is also
rounded up to 16-bytes alignment.

Fixes #29535

Signed-off-by: Shih-Wei Teng <swteng@andestech.com>

andrewboie commented:
what architecture is this a problem on?
the thing is, the initial frame is supposed to be an initial stack buffer context. so it seems very weird that the sizeof the initial frame is actually not already aligned properly, since it represents contents that will be popped off the stack at some point. can you provide some detail? is this for an arch that has an entry stub that moves stack contents to registers?

tyrantlizard77 reviewed (approved):
change looks fine BTW I can't ever see this hurting anything

andrewboie commented:
```
===================================================================
START - test_slice_reset
E: General protection fault (code 0x0)
E: RAX: 0x0000000000000000 RBX: 0x00000000001b0458 RCX: 0x000000000000000c RDX: 0x000000000000014a
E: RSI: 0x000000000014a000 RDI: 0x00000000001b0458 RBP: 0x000000000014dfe0 RSP: 0x00000000001b02f0
E:  R8: 0x0000000000000000  R9: 0x0000000000000000 R10: 0x0000000000000000 R11: 0x0000000000000000
E: R12: 0x0000000000000000 R13: 0x0000000000000000 R14: 0x0000000000101dcf R15: 0x0000000000000000
E: RSP: 0x00000000001b02f0 RFLAGS: 0x0000000000000202 CS: 0x0018 CR3: 0x00000000001bf000
E: RIP: 0x0000000000001133
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x0000000000121360 (unknown)
E: Halting system
```
are you able to reproduce this? seen in tests/kernel/sched/schedule_api on x86_64

tyrantlizard77 reviewed (changes requested):
I think this is causing x86_64 to do a mis-aligned `fxsave` or `fxrstor` operation.

would like some detail on which arch this fixes things for you.

katsuster commented:
This change breaks other arch, so for now is it better to apply only for RISC-V?

tyrantlizard77 reviewed (changes requested):
> @andrewboie
> RISC-V standard calling convention requires stack pointer must be 16bytes aligned.

OK, sure. That much is clear to me.

> If users use only C functions compilers keep SP 16bytes aligned, but in z_setup_new_thread(), it sets SP manually.
> So Zephyr violates SP alignment rules of RISC-V

The kernel sets the initial stack pointer to the very end of the entire stack object, which was defined with K_KERNEL_STACK_DEFINE or K_THREAD_STACK_DEFINE. The base address and size of the stack object should be at *least* aligned to ARCH_STACK_PTR_ALIGN, so the statement in setup_thread_stack():

```
	stack_ptr = (char *)stack + stack_obj_size;
```

Should be aligned.

Then we compute a `delta` value, which is also aligned:

```
	delta = ROUND_UP(delta, ARCH_STACK_PTR_ALIGN);
```

And the initial stack pointer set to the arch layer is just the difference, which should also be aligned:

```
	stack_ptr -= delta;
```

So you should never be getting an unaligned stack pointer sent to `arch_new_thread()` unless something is set up wrong.

Looking at arch.h in Risc-V we have:

```
#define ARCH_KERNEL_STACK_OBJ_ALIGN	Z_RISCV_PMP_ALIGN
#define ARCH_THREAD_STACK_SIZE_ADJUST(size) \
		Z_POW2_CEIL(ROUND_UP((size), Z_RISCV_PMP_ALIGN))
#define ARCH_THREAD_STACK_OBJ_ALIGN(size) \
		ARCH_THREAD_STACK_SIZE_ADJUST(size)
```

And that Z_RISCV_PMP_ALIGN is 4. That's your root cause. You should define these in terms of MAX(Z_RISCV_PMP_ALIGN, ARCH_STACK_PTR_ALIGN) and not just Z_RISCV_PMP_ALIGN.

The proposed patch I don't think is as robust as there could be other problems if the object itself is mis-aligned. I think it's better to ensure the object itself is aligned to the HW alignment.

Feel free to also amend ARCH_* macro documentation in arch_interface.h if the current description is unclear.

katsuster commented:
@andrewboie 
Thanks a lot for your clarification. I'm also checking kernel stack initialization behavior.
Current RISC-V port sets 16bytes aligned size to 'stack_obj_size' (= 1024).

Here is an example of result that is dumped stack pointer calculations:

```
===================================================================
START - test_pi
setup_thread_stack:472 stack_size:1024
setup_thread_stack:474 stack_obj_size:1024    <-- stack_obj_size is 16bytes aligned
setup_thread_stack:483 stack:0x80008fa0 stack_ptr:0x800093a0
setup_thread_stack:530 stack_ptr:0x800093a0 delta:0    <-- delta is 0, stack_ptr is 16bytes aligned
arch_new_thread:40 stack_ptr:0x800093a0
arch_new_thread:42 stack_init:0x800092fc    <-- here breaks 16bytes alignment rule

...

Pi calculation OK after 50 (high) + 2 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)

👆 Calculated result is correct, but printk() prints strange value because stack ptr is not 16bytes aligned.
```

The result shows that last value of stack_ptr at setup_thread_stack() is 16bytes aligned.

I understand and agree with you that you described current RISC-V port issue.
However it happens only when we enable CONFIG_PMP_STACK_GUARD, I think.

Especially in  a target  case of this PR, this is not root cause. Problem is in here:

```
void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
		     char *stack_ptr, k_thread_entry_t entry,
		     void *p1, void *p2, void *p3)

	stack_init = Z_STACK_PTR_TO_FRAME(struct __esf, stack_ptr);


// zephyr/include/sys/thread_stack.h

#define Z_STACK_PTR_TO_FRAME(type, ptr) \
	(type *)((ptr) - sizeof(type))
```

I think this is an issue for RISC-V, and we should fix it. But it seems first proposed PR brings bad effects on other architectures...
So I asked you "Just for now, is it better to apply only for RISC-V?".

Teng-Shih-Wei commented:
@andrewboie I agree with you that the bottom of allocated stack is forced to 16-bytes alignment in each step. Of course,  "Z_RISCV_PMP_ALIGN" is a weak point in latest Zephyr. I can create another PR for this enhancement. 

As @katsuster mentions, there is an problem if the exception stack frame is not 16-bytes aligned. In RISC-V port, it rounds up the exception stack frame to 16-bytes aligned in "arch/riscv/core/offsets/offsets.c"

GEN_ABSOLUTE_SYM(__z_arch_esf_t_SIZEOF, STACK_ROUND_UP(sizeof(z_arch_esf_t)));

Then the generated macro "__z_arch_esf_t_SIZEOF" is used for context switch in "arch/riscv/core/isr.S". Therefore, if we do not round up the exception stack frame in "arch_new_thread()", the stack point will exceed the bottom of the original allocated stack after thread context restored. It will pollute the memory outside the allocated stack after the thread runs. It may cause fatal problems in run time. Maybe we can apply it only for RISCV now, and have the better way for applying all architectures in the future. What do you think?

Teng-Shih-Wei committed:
RISC-V: Round up pre-populated stack frame to arch stack alignment

The stack frame size, used for context switch, is rounded up to 16-bytes
alignment. Therefore, we need round down the pointer of top of the
pre-populated stack frame so that the preserved stack frame size is also
rounded up to 16-bytes alignment.

Fixes #29535

Signed-off-by: Shih-Wei Teng <swteng@andestech.com>

andrewboie commented:
My review stands.

katsuster commented:
Reopened...
We need to merge this PR. Current RV32 Zephyr printf() outputs very strange values by this misalignment of stack frame.

@andrewboie
Sorry for asking again. I couldn't catch your concern points... I think @Teng-Shih-Wei just wants to adjust top of stack frame to be aligned with 16-bytes.

```
===================================================================
START - test_pi
Pi calculation OK after 50 (high) + 118 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)
Pi calculation OK after 150 (high) + 360 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)
Pi calculation OK after 250 (high) + 595 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)
Pi calculation OK after 350 (high) + 828 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)
Pi calculation OK after 450 (high) + 1054 (low) tests (computed -26815615863448970000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000)
 PASS - test_pi in 12.306 seconds
```

cfriedt commented:
@katsuster , @Teng-Shih-Wei 

With a second commit in this PR for testsuite coverage, it might make things a bit clearer. Please document the test case as well.

katsuster commented:
This PR is deprecated because #35831 has solved this problem.
So let's close.
